TGUI Outfit Manager and Editor

TheTrigg opened the pull request:
## About The Pull Request

The Outfit Manager has been replaced with a pretty TGUI interface.

Custom Outfits are now loaded and saved using SSpersistence.

I never really messed around with events and stuff when I was part of a downstream server, so it's hard for me to figure out what would be useful and what would just be too situational; so feedback is always useful.


## Why It's Good For The Game

Outfits for events can be created and tweaked much more easily, and with pretty UIs in the process.

https://user-images.githubusercontent.com/36010999/113460712-0d546180-93f0-11eb-80bd-aef0a0c58e63.mp4

https://user-images.githubusercontent.com/36010999/114125551-169c6d00-98cd-11eb-8dbd-f959a546f7f3.mp4

## Changelog
:cl: Trigg
admin: Admins just got a new TGUI Outfit Manager to mess around and make events with!
admin: Custom Outfits are now persistent and saved server-side, but can still be locally saved and loaded.
/:cl:

TheTrigg committed:
electric boogaloo
and this time I fixed my email

TheTrigg committed:
never been more unsure about how to lay out an UI before

TheTrigg committed:
hmmm

TheTrigg committed:
more d y n a m i c

TheTrigg committed:
holy shit it suddenly actually looks good now

TheTrigg committed:
rearrange

TheTrigg committed:
wack

TheTrigg committed:
still wack

TheTrigg committed:
naming innit

TheTrigg committed:
I always forget something

TheTrigg committed:
this is getting out of hand, now there are two of them

TheTrigg committed:
ui_act
kinda jank still tho

TheTrigg committed:
progress

TheTrigg committed:
ok fine it's time to rework glob.custom_outfits
if you don't see me on the other side you know what happened

TheTrigg committed:
actually this is much easier and much simpler
I'll probably leave GLOB.custom_outifts rework to some other PR,
then also make them persistant using the DB probably

TheTrigg committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into slightly-smater-trigg-is-at-it-again-with-tgui

TheTrigg committed:
clean up old manager, maybe PR ready?

TheTrigg committed:
proper var naming
see I've learned my lesson
mostly

TriggeredBoi commented on code/modules/admin/outfit_editor.dm:
```diff
@@ -0,0 +1,187 @@
+	clear_human_dummy(dummy_key)
+	qdel(src)
+
+/datum/outfit_editor/proc/init_dummy()
```
This is the second time I've used dummies in pretty much the same way, maybe I should make some centralized way of using dummy lookalikes. Not in this PR though.
By the way I haven't forgotten about the `equip` problem pointed out in https://github.com/tgstation/tgstation/issues/57865

TriggeredBoi commented on code/modules/admin/outfits.dm:
```diff
@@ -1,35 +1,14 @@
-	dat += "<a href='?_src_=holder;[HrefToken()];load_outfit=1'>Load from file</a>"
-	admin << browse(dat.Join(),"window=outfitmanager")
 
 /datum/admins/proc/save_outfit(mob/admin,datum/outfit/O)
```
Why are these procs under `/datum/admins`? Can/should I move them?

Kylerace commented on code/datums/outfit.dm:
```diff
@@ -347,6 +347,35 @@
+	OUTFIT_COPY(backpack_contents)
+	OUTFIT_COPY(box)
+	OUTFIT_COPY(implants)
+	OUTFIT_COPY(accessory)
```
not really a need to have this define, it doesnt save on much typing and is arguably less clear

TriggeredBoi commented:
Ah fudge I forgot to make it a draft sorry

TheTrigg committed:
bad define is gone
tbh I didn't like it either

TheTrigg committed:
outfits now have a dropdown shortcut to the editor in VV

TheTrigg committed:
ctrl-clicking
was gonna be shift-clicking but it was weirdly jank

tralezab commented on tgui/packages/tgui/interfaces/OutfitEditor.js:
```diff
@@ -0,0 +1,138 @@
+
+export const OutfitEditor = (props, context) => {
+  const { act, data } = useBackend(context);
+  const { outfit, saveable } = data;
```
dummy64 should be in this data, in a few spots i see you using data.thing but that's bad for readability

tralezab commented on tgui/packages/tgui/interfaces/OutfitEditor.js:
```diff
@@ -0,0 +1,138 @@
+      width={380}
+      height={625}>
+      <Window.Content>
+
```
remove this space

tralezab commented on tgui/packages/tgui/interfaces/OutfitEditor.js:
```diff
@@ -0,0 +1,138 @@
+          style={{
+            '-ms-interpolation-mode': 'nearest-neighbor',
+          }} />
+
```
remove this space

tralezab commented on tgui/packages/tgui/interfaces/OutfitEditor.js:
```diff
@@ -0,0 +1,138 @@
+                tooltip="Edit this outfit on a VV window"
+                tooltipPosition="left"
+                onClick={() => act("vv", {})}
+              />
```
move this to the line above

tralezab commented on tgui/packages/tgui/interfaces/OutfitEditor.js:
```diff
@@ -0,0 +1,138 @@
+                tooltipPosition="left"
+                title={!saveable && "This outfit is already on the custom outfit list. Any changes made here will be immediately applied."}
+                onClick={() => act("save", {})}
+              />
```
move this to the line above

tralezab commented on tgui/packages/tgui/interfaces/OutfitEditor.js:
```diff
@@ -0,0 +1,138 @@
+              />
+            </>
+          }>
+
```
remove this space

tralezab commented on tgui/packages/tgui/interfaces/OutfitEditor.js:
```diff
@@ -0,0 +1,138 @@
+          </Box>
+
+        </Section>
+
```
two needless spaces here

tralezab commented on tgui/packages/tgui/interfaces/OutfitEditor.js:
```diff
@@ -0,0 +1,138 @@
+          e.ctrlKey
+            ?act("ctrlClick", { slot: slot })
+            :act("click", { slot: slot }); }}
+      >
```
move this to the line above, also cool use of ctrlKey

tralezab commented on tgui/packages/tgui/interfaces/OutfitEditor.js:
```diff
@@ -0,0 +1,138 @@
+const OutfitSlot = (props, context) => {
+  const { act, data } = useBackend(context);
+  const { name, icon, slot } = props;
+  const currItem = data.outfit[slot];
```
this should just be in a { } = data

tralezab commented on tgui/packages/tgui/interfaces/OutfitManager.js:
```diff
@@ -0,0 +1,90 @@
+      width={300}
+      height={300}>
+      <Window.Content>
+
```
remove this space

tralezab commented on tgui/packages/tgui/interfaces/OutfitManager.js:
```diff
@@ -0,0 +1,90 @@
+            ))}
+          </Stack>
+        </Section>
+
```
remove this space

tralezab commented on tgui/packages/tgui/interfaces/OutfitEditor.js:
```diff
@@ -0,0 +1,138 @@
+              <OutfitSlot name="Shoes" icon="socks" slot="shoes" />
+              <OutfitSlot name="Left Pocket" slot="l_pocket" />
+              <OutfitSlot name="Right Pocket" slot="r_pocket" />
+            </Stack>
```
this code above here is pretty dope through and through, i hope you give each slot an icon like you did with the shoes

TriggeredBoi replied:
Thanks - and absolutely, that's the plan.

tralezab reviewed (commented):
here's what i would change (or just a few compliments on how you did bits and bobs, all in all this ui is super innovative and really cool

TheTrigg committed:
review
much, much better than last time

TheTrigg committed:
couple icons, don't know what do do with the other ones
smh why doesn't fontawesome always have exactly what I need reee

Pumpkinoe commented:
this is so cool

AnturK commented:
Instead of jsons you could just dump these into local storage now. Would need changes to validation.

TriggeredBoi commented:
> Instead of jsons you could just dump these into local storage now. Would need changes to validation.

uhhhhhhh how?

AnturK commented:
Sorry for late reply, you can take a look at [pod launcher interface](https://github.com/tgstation/tgstation/blob/master/tgui/packages/tgui/interfaces/CentcomPodLauncher.js#L760) for quick example. Mind you'll have to validate if paths stored in the storage are still up to date when using them again.

TheTrigg committed:
better code, background bounding fixed

TheTrigg committed:
really out of ideas for icons

TriggeredBoi commented:
> Sorry for late reply, you can take a look at [pod launcher interface](https://github.com/tgstation/tgstation/blob/master/tgui/packages/tgui/interfaces/CentcomPodLauncher.js#L760) for quick example. Mind you'll have to validate if paths stored in the storage are still up to date when using them again.

After banging my head against that example and finally figuring out how it worked, I don't believe `storage` is the best option for storing outfits because:
1 - Outfits would not be shared between admins (but I still really like the idea of them carrying over between servers)
2 - While I can see how I could maybe do it, it wouldn't be the prettiest of implementations as I pack more than just the typepath of each slot to the OutfitEditor [in ui_data](https://github.com/tgstation/tgstation/pull/58125/files#diff-01286cdbd93c2a09be2c0549fb6f7de9392c0cd78a4bf2dcfe3236a5cf18d654R55)

As an alternative, however, I plan on figuring out how the DB stuff works tomorrow and maybe store them there instead... Or would that be too overkill?

TheTrigg committed:
cool
it works, just not _properly_

TheTrigg committed:
works properly, but every outfit is saved in a single json
this way I don't need to worry about outfits with duplicate names
the implementation is also much simpler and readable

TheTrigg committed:
I always forget something before committing

TheTrigg committed:
sigh

TheTrigg committed:
CI please

TriggeredBoi commented:
Oh hey, it worked. Even though it took me the entire day to do it, it was still easier than I thought it would be. Cool.
So GLOB.outfits is saved to a json on roundend now, and outfits with duplicate names are still allowed.

I still have a couple loose ends to fix up, after that this PR will probably be ready for review (I already feel bad for poor man who gets to review my still trash DM code)

TheTrigg committed:
tiny redesign

TheTrigg committed:
pocket and hand icons

TheTrigg committed:
adjusted tooltips, clear option

TheTrigg committed:
last touches before review ready

TheTrigg committed:
organize imports

TriggeredBoi commented:
It's time. I feel bad for the reviewer, something tells me there will be a lot to address.

stylemistake committed:
Update outfit_manager.dm

stylemistake committed:
Update OutfitManager.js

stylemistake committed:
Update outfit_editor.dm

stylemistake committed:
Update OutfitEditor.js

TriggeredBoi commented:
...Or not
